json: Fix multidimensional array support

mrfuchs opened the pull request:
This patch fixes support for encoding and decoding multidimensional arrays
as described by the JSON_OBJ_DESCR_ARRAY_ARRAY() macro.

Fixes https://github.com/zephyrproject-rtos/zephyr/issues/50801

mrfuchs committed:
json: Fix multidimensional array support

This patch fixes support for encoding and decoding multidimensional arrays
as described by the JSON_OBJ_DESCR_ARRAY_ARRAY() macro.

Currently, the JSON array encoding and decoding functions, arr_encode()
and arr_parse(), expect array elements to be of object or primitive type.
However, arrays may be nested and so an array's elements may also be
arrays.

In order to support nested arrays, two special cases must be considered:

1. The array of objects/arrays sub-descriptor is described by two
`json_obj_descr` structs and so two instead of one `json_obj_descr`
structs must be skipped when iterating over the JSON descriptor to get to
an array's elements.
2. The implicit array item count field has to be considered for the
parent itself and all its child array items when calculating an element's
size.

Fixes #50801

Signed-off-by: Markus Fuchs <markus.fuchs@ch.sauter-bc.com>

mrfuchs committed:
tests: lib: json: Fix and enhance test for array of arrays

Fix encoded JSON string in test_json_decoding_array_array() test so it
matches the described array object and add a test case for encoding
arrays of arrays.

Signed-off-by: Markus Fuchs <markus.fuchs@ch.sauter-bc.com>

mrfuchs committed:
tests: lib: json: Add tests for handling two-dimensional arrays

Add tests for encoding and decoding two-dimensional arrays as described by
the JSON_OBJ_DESCR_ARRAY_ARRAY() macro.

Signed-off-by: Markus Fuchs <markus.fuchs@ch.sauter-bc.com>

carlescufi commented:
@mrfuchs can you take a look at the CI failures?

andyross commented on tests/lib/json/src/main.c:
```diff
@@ -450,6 +486,130 @@ ZTEST(lib_json_test, test_json_obj_arr_decoding)
+			[0] = {
+				.elements = {
+					[0] = { .name = "Simón Bolívar",   .height = 168 },
+					[1] = { .name = "Pelé",            .height = 173 },
```
What's our policy on UTF-8 in source code?  IIRC that's a C11 feature, right?

stephanosio replied:
There is no policy on the usage of Unicode characters in source files, as far as I know (cc @tejlmand @nashif).

The support for non-ASCII characters in source files is implementation defined; but, I do not think there are any modern compilers (as in, released after 2000) that outright reject the source files that contain non-ASCII characters.

As for the UTF-8 encoded characters in the string literals, as far as I can see, they are never put through any locale-aware functions, and the validation inputs and outputs are in the same file (i.e. must use the same encoding), this should not be a problem in itself.

keith-packard replied:
It looks like C is slowly moving to accepting utf-8 in source code. As for source code, gcc has the `-finput-charset` and `-fexec-charset` command line options, which should probably be set to `UTF-8` to avoid problems building in locales using a different charset, in which case it's hard to know exactly what the compiler would do with utf-8 sequences in strings. I recall having adventures when building Java applications on Windows as the default charset there was UTF-16. Not sure what C compilers do in the Windows world.

keith-packard replied:
If you want to avoid all issues, you would need to use octal escaped sequences for utf-8 in strings. For tests, that's probably a reasonable plan.

mrfuchs replied:
Updated the unit tests accordingly.

andyross reviewed (dismissed):
Saw this languishing and figured I should give it an approval, even though it's not really my area (we don't really have a maintainer for JSON right now, I guess?).

Language wonks (@stephanosio @keith-packard ?) might care about the unicode string literals, though they seem harmless to me.

mrfuchs committed:
tests: lib: json: Add test for array of objects

Add tests for encoding and decoding nested arrays of objects located
inside a parent object at a non-zero offset.

Signed-off-by: Markus Fuchs <markus.fuchs@ch.sauter-bc.com>

mrfuchs committed:
json: Fix 64-bit support

This patch fixes encoding arrays of objects on 64-bit targets.

Fixes #36696

Signed-off-by: Markus Fuchs <markus.fuchs@ch.sauter-bc.com>

dariis commented:
@mrfuchs Thank you, it worked for me :+1:

mrfuchs commented:
> Saw this languishing and figured I should give it an approval, even though it's not really my area

Thank you! Unfortunately, I had to update this PR. Would you review it once more, please?

> (we don't really have a maintainer for JSON right now, I guess?).

Yes, it's difficult to get JSON PRs approved - my last PR #47989 also got closed due to a lack of reviewers.
@Sir-Branch asked for being the official JSON maintainer, but has become inactive lately. Sometimes @martinjaeger helps reviewing changes to the JSON library.

> Language wonks (@stephanosio @keith-packard ?) might care about the unicode string literals, though they seem harmless to me.

I didn't really pay attention to that. I just reused the test data that was already there.

mrfuchs commented:
I noticed that my original patch would still not fix parsing arrays of objects. That's why I moved these lines inside the loop...

https://github.com/zephyrproject-rtos/zephyr/blob/55dc3986060ab3ef29aeeee3ba168bd65bdb23ba/lib/os/json.c#L578-L583

... and added a test for it (cf96b83). However, after adding the new unit test, CI failed again (cf. #36696 and [#24372#issuecomment-617392567](https://github.com/zephyrproject-rtos/zephyr/issues/24372#issuecomment-617392567)). So, I fixed this issue in 55dc398 to make CI work again.

mrfuchs commented:
Reopening due to #58493.

keith-packard reviewed (changes requested):
Please switch the non-ascii text in the tests to use octal escape sequences to avoid potential compiler issues on platforms not defaulting to utf-8 text encodings.

mrfuchs commented:
> I would like to inquire about the current status of this fix as I am experiencing the same issue [...]

Will update this PR soon to follow @keith-packard's suggestion re. non-ascii text in the unit tests.

mrfuchs committed:
tests: lib: json: Switch non-ascii text to octal escape sequences

Switch the non-ascii text in the tests to use octal escape sequences
to avoid potential compiler issues on platforms not defaulting to
utf-8 text encodings.

Signed-off-by: Markus Fuchs <markus.fuchs@ch.sauter-bc.com>

mrfuchs commented:
@andyross, @keith-packard, I updated the unit tests. Could you review? Thanks!

keith-packard reviewed (commented):
The commit message on the first patch (fixing the bugs) provides no information about what was broken and how the change addresses that. Makes reviewing the patch pretty hard.

keith-packard commented:
> @andyross, @keith-packard, I updated the unit tests. Could you review? Thanks!

Unit test changes to use octal escapes look good!
